### PR TITLE
fix: restore original ZDOTDIR after activation

### PR DESCRIPTION
**Pull request by limeytexan**
## Proposed Changes
    
This patch refactors the flox-provided `zdotdir` collection of zsh init scripts to ensure that the original value of ZDOTDIR is restored following activation, and amends the activation logic to be more consistent and robust in interactive login shell case (which previously could be broken by a user's ~/.login file).

This patch also updates the README.md file to reflect the updated logic.
    
Finally, this patch updates the trace verbosity for zsh to selectively enable `compinit` tracing only with verbosity levels 3 and above, whereas it had previously been enabled at level 2. This tracing is particularly verbose, which had significantly reduced the usefulness of the tracing.

## Release Notes

* fixed bug (#1463) which caused the user-specified `ZDOTDIR` to be unset with flox activations

### CODE REVIEW

File: cli/tests/activate.bats [hunk 1]
```diff
@@ -924,3 +924,23 @@ EOF
 }
 
 # ---------------------------------------------------------------------------- #
+
+# bats test_tags=activate,activate:zdotdir,activate:zdotdir:bash
+@test "bash: preserve custom ZDOTDIR" {
```
Comment (mkenigs) [inline cli/tests/activate.bats:929]: question nonblocking: 🤔 do we touch ZDOTDIR for a bash activation?

**Review by mkenigs — CHANGES_REQUESTED** [review]
blocking: the `[[ -o interactive ]] || [[ -o login ]] || [ -z "$FLOX_ZSH_INIT_SCRIPT" ] || source "$FLOX_ZSH_INIT_SCRIPT"` changes need tests or to be in a separate PR

File: cli/tests/activate.bats (hunk 1 above)
Comment (limeytexan) [reply cli/tests/activate.bats:929]: We don't, but thought it seemed a reasonable test to ensure we don't inadvertently set the variable - you never know when someone will launch zsh as a child process of a bash invocation.

**Comment by limeytexan** [comment]
> blocking: the [[ -o interactive ]] || [[ -o login ]] || [ -z "$FLOX_ZSH_INIT_SCRIPT" ] || source "$FLOX_ZSH_INIT_SCRIPT" changes need tests or to be in a separate PR

@mkenigs can you clarify? The updates to README.md explain the reasoning behind the change - are you explicitly requesting for a test to be added covering the zsh login shell case?

**Comment by mkenigs** [comment]
> > blocking: the [[ -o interactive ]] || [[ -o login ]] || [ -z "$FLOX_ZSH_INIT_SCRIPT" ] || source "$FLOX_ZSH_INIT_SCRIPT" changes need tests or to be in a separate PR
> 
> @mkenigs can you clarify? The updates to README.md explain the reasoning behind the change - are you explicitly requesting for a test to be added covering the zsh login shell case?

Yes - the logic is spread across multiple files and depends on knowledge of how zsh initialization works. It could completely break `activate` in certain cases if the logic is messed up. I think that either needs a test or a good explanation of why we can't test it

File: cli/tests/activate.bats [hunk 2]
```diff
@@ -924,3 +924,76 @@ EOF
...
+# ---------------------------------------------------------------------------- #
+
+# bats test_tags=activate,activate:zdotdir,activate:zdotdir:zshenv
+@test "zdotdir: test zshenv activation" {
+  echo "echo sourcing .zshenv" > "$HOME/.zshenv"
+  echo "echo sourcing .zshrc" > "$HOME/.zshrc"
+  echo "echo sourcing .zlogin" > "$HOME/.zlogin"
+  "$FLOX_BIN" delete -f
+  "$FLOX_BIN" init
+  run "$FLOX_BIN" install hello
```
Comment (mkenigs) [inline cli/tests/activate.bats:955]: nonblocking: do we need the `install`? That just makes the test slower

**Comment by limeytexan** [comment]
Spoke with @mkenigs offline, added tests for each of the [3] zsh invocation cases: non-interactive, interactive non-login, and interactive login. Each of these verify the correct operation of the flox zsh init script by exercising the ability to invoke `hello` as installed to the environment.

**Review by mkenigs — APPROVED** [review]

File: cli/tests/activate.bats (hunk 2 above)
Comment (limeytexan) [reply cli/tests/activate.bats:955]: yeah I was looking for some other way to do that; there's no sentinel variable set by the zsh init script, the best I can imagine is looking for `$FLOX_ENV/bin` as the first element in the `$PATH`. If you want to make that change then it's fine by me, or we could add a sentinel variable to each of the `./pkgdb/src/buildenv/assets/activate.d/{bash,zsh}` init scripts?

Comment (limeytexan) [reply cli/tests/activate.bats:955]: Added patch to fix this in https://github.com/flox/flox/pull/1475

**Pull request merged**